json_rpc/servers/httpserver.nim

```diff
@@ -1,233 +1,64 @@
-  while true:
-    try:
-      let hlenfut = transp.readUntil(addr buffer[0], MaxHttpHeadersSize,
-                                     HeadersMark)
-      let ores = await withTimeout(hlenfut, HttpHeadersTimeout)
-      if not ores:
-        # Timeout
-        debug "Timeout expired while receiving headers",
-              address = transp.remoteAddress()
-        discard await transp.sendAnswer(HttpVersion11, Http408)
-        await transp.closeWait()
-        break
-      else:
-        let hlen = hlenfut.read()
-        buffer.setLen(hlen)
-        header = buffer.parseRequest()
-        if header.failed():
-          # Header could not be parsed
-          debug "Malformed header received",
-                address = transp.remoteAddress()
-          discard await transp.sendAnswer(HttpVersion11, Http400)
-          await transp.closeWait()
-          break
-    except TransportLimitError:
-      # size of headers exceeds `MaxHttpHeadersSize`
-      debug "Maximum size of headers limit reached",
-            address = transp.remoteAddress()
-      discard await transp.sendAnswer(HttpVersion11, Http413)
-      await transp.closeWait()
-      break
-    except TransportIncompleteError:
-      # remote peer disconnected
-      debug "Remote peer disconnected", address = transp.remoteAddress()
-      await transp.closeWait()
-      break
-    except TransportOsError as exc:
-      debug "Problems with networking", address = transp.remoteAddress(),
-            error = exc.msg
-      await transp.closeWait()
-      break
-    except CatchableError as exc:
-      debug "Unknown exception", address = transp.remoteAddress(),
-            error = exc.msg
-      await transp.closeWait()
-      break
-
-    let vres = await validateRequest(transp, header)
-
-    if vres == Success:
-      trace "Received valid RPC request", address = $transp.remoteAddress()
-
-      # we need to get `Connection` header value before, because
-      # we are reusing `buffer`, and its value will be lost.
-      connection = header["Connection"]
-
-      let length = header.contentLength()
-      buffer.setLen(length)
-      try:
-        let blenfut = transp.readExactly(addr buffer[0], length)
-        let ores = await withTimeout(blenfut, HttpBodyTimeout)
-        if not ores:
-          # Timeout
-          debug "Timeout expired while receiving request body",
-                address = transp.remoteAddress()
-          discard await transp.sendAnswer(header.version, Http413)
-          await transp.closeWait()
-          break
+    httpServers: seq[HttpServerRef]
+
+proc addHttpServer*(rpcServer: RpcHttpServer, address: TransportAddress) =
+  proc processClientRpc(rpcServer: RpcHttpServer): HttpProcessCallback {.closure.} =
+    return proc (req: RequestFence): Future[HttpResponseRef] {.async.} =
+      if req.isOk():
+        let request = req.get()
+        let body = await request.getBody()
+
+        let future = rpcServer.route(cast[string](body))
```

Comment:
The `stew` library have some safe helper functions for converting bytes to strings:

https://github.com/status-im/nim-stew/blob/master/stew/byteutils.nim#L154

Reply:
73584e8e23cdc6a8b010e8123db0b5dfa7def720